File: backend/app/schemas/suggestions.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.models import SuggestionStatus, SuggestionType


class ChunkEvidenceItem(BaseModel):
    """Evidencia de un chunk individual mostrada al revisor."""

    chunk_id: str
    content: str
    chunk_index: int
    token_count: int
    page_number: int | None = None


class SuggestionResponse(BaseModel):
    id: uuid.UUID
    document_id: uuid.UUID
    type: SuggestionType
    status: SuggestionStatus
    description: str
    reasoning: str | None
    confidence_score: float
    source_chunk_ids: list
    source_doc_id: str
    source_web_url: str | None = None
    source_type: str | None = None
    review_reason: str | None
    reviewed_by: uuid.UUID | None
    # HU-26 — identidad legible de quien revisó la sugerencia
    reviewed_by_email: str | None = None
    reviewed_by_name: str | None = None
    reviewed_at: datetime | None
    created_at: datetime
    document_name: str | None = None
    source_chunks: list[ChunkEvidenceItem] = []

    model_config = {"from_attributes": True}


class SuggestionsListResponse(BaseModel):
    items: list[SuggestionResponse]
    total: int


class ApproveResponse(BaseModel):
    id: uuid.UUID
    status: SuggestionStatus
    message: str


class RejectRequest(BaseModel):
    reason: str = Field(..., min_length=1, description="Motivo del rechazo")


class RejectResponse(BaseModel):
    id: uuid.UUID
    status: SuggestionStatus
    message: str


class FeedbackRequest(BaseModel):
    comment: str | None = None


class DocumentHistoryEntry(BaseModel):
    id: uuid.UUID
    doc_id: uuid.UUID
    action: str
    performed_by: uuid.UUID | None
    timestamp: datetime
    before_content: dict | None
    after_content: dict | None
    reason: str | None

    model_config = {"from_attributes": True}


class AnalyticsResponse(BaseModel):
    total_documents: int
    by_status: dict[str, int]
    total_suggestions: int
    suggestions_by_status: dict[str, int]
    suggestions_by_type: dict[str, int]
    approval_rate: float

File: frontend/src/api/suggestions.ts
import api from "./axios";

export type SuggestionType =
  "redundancy" | "conflict" | "faq" | "update" | "inconsistency";
export type SuggestionStatus = "pending" | "approved" | "rejected";

// Severidad para hallazgos de inconsistencia
export type SeverityLevel = "high" | "medium" | "low";

export interface InconsistencyEvidence {
  chunk_id_a: string;
  chunk_id_b: string;
  doc_id_a: string;
  doc_id_b: string;
  extract_a: string;
  extract_b: string;
  description: string;
  suggestion: string;
  severity: SeverityLevel;
  type: "self_contradiction" | "terminology" | "numerical" | "structural";
}

export interface ChunkEvidence {
  chunk_id: string;
  content: string;
  chunk_index: number;
  token_count: number;
  page_number: number | null;
}

export interface Suggestion {
  id: string;
  document_id: string;
  document_name: string | null;
  type: SuggestionType;
  status: SuggestionStatus;
  description: string;
  reasoning: string | null;
  confidence_score: number;
  source_chunk_ids: string[];
  source_chunks: ChunkEvidence[];
  source_doc_id: string;
  source_web_url: string | null;
  source_type: string | null;
  review_reason: string | null;
  reviewed_by: string | null;
  /** HU-26 — identidad legible del revisor */
  reviewed_by_email?: string | null;
  reviewed_by_name?: string | null;
  created_at: string;
  reviewed_at: string | null;
}

export interface SuggestionsListResponse {
  items: Suggestion[];
  total: number;
}

export interface AnalyticsData {
  total_documents: number;
  by_status: Record<string, number>;
  total_suggestions: number;
  suggestions_by_status: Record<string, number>;
  suggestions_by_type: Record<string, number>;
  approval_rate: number;
}

export const getSuggestions = (params?: {
  status?: string;
  type?: string;
  doc_id?: string;
  page?: number;
  limit?: number;
}) => api.get<SuggestionsListResponse>("/api/suggestions", { params });

export const approveSuggestion = (id: string) =>
  api.post<Suggestion>(`/api/suggestions/${id}/approve`);

export const rejectSuggestion = (id: string, reason: string) =>
  api.post<Suggestion>(`/api/suggestions/${id}/reject`, { reason });

export const addFeedback = (id: string, comment?: string) =>
  api.post(`/api/suggestions/${id}/feedback`, { comment });

export const getAnalytics = () => api.get<AnalyticsData>("/api/analytics");
